### DISCUSSION TITLE
Dovecot - Mail Compression Plugin question

### DISCUSSION THREAD

**Discussion by dracossan** [docker-mailserver/docker-mailserver#3664] [Q&A]
Hi,

First of all, thanks to the dev and the community on this project. Using it since around a year, and this has been so helpful to setup email server for people who don't have a PhD degrees in Email configuration 😄 . Thanks you.

Not sure it's the proper place, put I have a question on the dovecot [Mail Compression Plugin](https://doc.dovecot.org/3.0/configuration_manual/mail_compress_plugin/#mail-compress-plugin).

Tried to implement this features/option in a dev environment with no success (Docker Version 12.1.0).

Does the image contain all the sufficient requirement to be able to implement this plugin ? 

Does it 'safe' to implement this option as it's seems to have some security concerned ? (read on some area).

Does anyone try or success to implement it and willing to share some documentation ?

and maybe last one more has a dev request, could it be possible to add it and be able to enabled it with some ENV variables ?

Thanks

**Comment by polarathene** [comment]
I haven't got time to look into this much for you, but you absolutely can open a feature request for it.

Although it's rare that someone contributes a feature request from someone else, you might get lucky!

---

> First of all, thanks to the dev and the community on this project.

Thanks! ❤️ We work very hard to maintain the project :)

> Does the image contain all the sufficient requirement to be able to implement this plugin?

Our packages are [managed here](https://github.com/docker-mailserver/docker-mailserver/blob/68a43eb4970f2ab7680ffa470a1d26e37fa375f0/target/scripts/build/packages.sh) which is used by the [`Dockerfile`](https://github.com/docker-mailserver/docker-mailserver/blob/68a43eb4970f2ab7680ffa470a1d26e37fa375f0/Dockerfile) that does some other setup.

`zstd` is presently missing, so that example will probably fail. You could try with a [different compressor](https://doc.dovecot.org/3.0/settings/plugin/mail-compress-plugin/#settings).

A future release of DMS will likely include `zstd`. If you'd like it, you could probably add `apt install zstd` to a [`user-patches.sh` script](https://docker-mailserver.github.io/docker-mailserver/v13.0/config/advanced/override-defaults/user-patches/), which would install the package each time the container starts.

---

> Is it 'safe' to implement this option as it's seems to have some security concerned? (read on some area).

I can't comment much on that due to lack of time. It seems to describe it as a precaution Dovecot takes to be safe, rather than a concern for you to worry about?:

> _To avoid IMAP clients attempting to exploit security holes in the compression algorithm libraries (e.g., bzlib) by writing specially crafted mails using IMAP’s APPEND command, Dovecot will not allow clients to save mails that are detected as compressed._

---

> Has anyone tried or had success to implement it and willing to share some documentation?
> could it be possible to add it and be able to enabled it with some ENV variables?

You could use `userpatches.sh` for any modifications needed. I'd advise it even for copying from your config volume mount to the internal location `/etc/dovecot/conf.d/` rather than mounting directly there.

Here's an example of a similar Dovecot plugin being configured in DMS:

https://github.com/docker-mailserver/docker-mailserver/blob/68a43eb4970f2ab7680ffa470a1d26e37fa375f0/target/scripts/startup/setup.d/dovecot.sh#L103-L111

So you might have this `user-patches.sh`:

```bash
#!/bin/bash

# Copy the file from config volume to the internal Dovecot config folder:
cp /tmp/docker-mailserver/custom/dovecot/90-compress.conf /etc/dovecot/conf.d/90-compress.conf

# Adds the `mail_compress` plugin to the existing plugins config:
sed -i \
     "s|mail_plugins = \$mail_plugins|mail_plugins = \$mail_plugins mail_compress|g" \
     /etc/dovecot/conf.d/10-mail.conf
```

This is a basic example, so it'll append the plugin to the [`10-mail.conf` line](https://github.com/docker-mailserver/docker-mailserver/blob/68a43eb4970f2ab7680ffa470a1d26e37fa375f0/target/dovecot/10-mail.conf#L209) each time it's run, so it's important to [(re)start DMS correctly](https://docker-mailserver.github.io/docker-mailserver/v13.0/config/debugging/#preliminary-checks) which would avoid that issue.

I'm not 100% sure if that's the correct line to update, or if you'd need to update the imap config too/instead. You'll need to experiment 👍

**Reply by dracossan** [reply]
@polarathene Thanks you. i will give it a try from my side following your recommendation !
Will revert if i have any issue.

**Comment by mbethke** [comment]
JFTR, I got compression working with the `zlib` plugin that has zstd linked in even on the current DMS. The `mail_compression` plugin seems to be a Dovecot 3 thing so I didn't look into it but I'd be surprised if it didn't work the same way. Invoking an external compressor for each little mail file doesn't sound like a good idea on a busy system.

My `user-patches.sh`:
```
#!/bin/sh
# ...
cat >/etc/dovecot/conf.d/10-mailcompress.conf <<EOF
mail_plugins = \$mail_plugins zlib

plugin {
  zlib_save = zstd
  zlib_save_level = 20
}
EOF
```
